teps/0089-nonfalsifiable-provenance-support.md

```diff
@@ -224,24 +291,95 @@ The entrypointer image should also be able to see that this flag is set and acco
+1. Verify that a necessary fields are present in the pod status result
+1. Verify that a TaskRun pod status being modified with incorrect results isn't verified by the TektonContoller
 
 Tests for Signed TaskRuns:
-1. Verify that a TaskRun that's been modified during execution isn't verified
-1. Verify that a TaskRun that's been modified after execution isn't verified
-
+1. Verify that a TaskRun that's been modified during execution isn't verified by the TektonController
+1. Verify that a TaskRun that's been modified after execution isn't verified by the TektonController
+1. Verify that a TaskRun that's been modified during execution isn't verified by the Chains
+1. Verify that a TaskRun that's been modified after execution isn't verified by the Chains
+
+## Threat Model
+
+There are 5 main components in this threat model:
+- TektonController (Signer/Verifier)
+- Chains (Verifier)
+- TaskRun Pod/Entrypointer (Signer)
+- SPIRE Server (Authority)
+- SPIRE Agent (Workload Attestor)
+
+There are several capabilities that we want to ensure are correct:
+- Verification of signed Results are signed by the correct TaskRun Pod/Entrypointer step
+  - Verified by TektonController through SPIRE Authority
+- Verification of signed TaskRuns are signed by the TektonController
+  - Verified by TektonController and Chains through SPIRE Authority
+
+Due to the implementation of these different components being similar in nature of how they interact with the architecture (k8s, SPIRE, etc.), the two cases can be evaluated against the same threats.
+
+Threats vectors around verification fall into several categories:
+- Ability to create a verifiable false signature
+- Able to influence a skip of verification step
+- Able to influence verification to verify against a false authority
+
+### Ability to create a verifiable false signature
+
+Potential threats:
+- Access to SPIRE upstream authority lets an attacker sign arbitrary values.
+- Access to pod execution environment allows minting false signatures for that TaskRun step
+  - Through exec'ing into a pod (via api-server).
+  - Vertical attack from host.
+- Ability to trick the minting of a false identity, or underspecified measurement/verification of identity (two signing entities are the same which should not be).
+
+
+Potential mitigations:
+- To prevent upstream authority from being breached, the SPIRE server should be located external to the cluster, and use a non-file upstream authority plugin (i.e. vault, gcp_cas, etc.).
+- To prevent pod execution environment access:
+  - k8s should be configured to disallow exec'ing into pods.
+  - Underlying host should be hardened and protected.
+  - Memory introspection features should be disabled to prevent memory introspection of keys.
+  - Disallow ptrace in pods.
+- Ensure the task runs created are uniquely identifiable. SPIFFE ID used should be uniquely idenfiable. e.g. Pod IDs can be included in TaskRun identifiers.
```

Comment:
hm this would be a big change for tekton - we allow users to name taskruns and while they are unique, you can delete an existing taskrun and create it again

do we need to use the taskruns to create the SPIFFE IDs or can we rely on just the pod names?

Reply:
> but just adding the pod IDs into the TaskRun status (and the SPIFFE IDs)

ah okay, if i'm understanding you correctly we've already got the pod ID in the taskrun status (https://github.com/tektoncd/pipeline/blob/main/docs/taskruns.md#monitoring-execution-status) (though we have talked about removing them - https://github.com/tektoncd/pipeline/issues/1689 - maybe this is a good reason not to)

> because annotations are updated separately

could you explain a bit more about what you mean that annotations are updated separately?

> While this feature is in alpha though, I think it is possible to house all these in the ["Status" of the TaskRunStatus](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/taskrun_types.go#L111). Thoughts on this?

interesting idea! in general my assumption would be that annotations are a better place for more experimental features b/c they can be changed without the API types needing to change - breaking API changes (even if our policy allows for them) mean potentially making existing resources invalid, e.g. a TaskRun that was created with a field that we later remove.

at the same time i wonder if the status of the taskrun is a better long term home for these fields. the main downside I can think of is that the value of these fields is very specific not just to k8s but to SPIRE in particular - we try to [limit the k8s specific API fields we add](https://github.com/tektoncd/community/blob/main/design-principles.md#conformance) (to theoretically support a non-k8s backed Tekton implementation, which is pretty much theoretical), and ideally we'd want to make sure that potentially we could swap SPIRE out for something else (if there was ever a comparable alternative which afaik there isnt XD)

Reply:
RE: pod names/uid. I think this is something worth exploring a bit more. It may not necessarily need to be part of an object, but just that the controller would need to have a way to distinctly verify the pod. It also does lean into how taskruns end up handling pods (which I am unclear on) - if the pod is killed and a new pod is spun up in its place, what happens here. Maybe owner reference is sufficient to manage this here, but not an expert on the inner workings on that. 

> at the same time i wonder if the status of the taskrun is a better long term home for these fields. 

Yea - the dream is that kubernetes would eventually make some of these mechanisms part of the API server. We would still be a while aways from this though.. 

> to theoretically support a non-k8s backed Tekton implementation, which is pretty much theoretical

I resound with this statement very very much.... support and backward compatibility is a pain especially with the velocity of k8s changes.

Reply:
> if the pod is killed and a new pod is spun up in its place, what happens here

im actually not sure what happens if you do that X'D i THINK the controller might start a new pod but im not certain - if it did start a new pod, i'm not sure it would start it from the beginning or it would be in some weird state. it also might consider it cancelled - imo if that happened the taskrun should stop and consider itself failed

what we should allow is a pod being rescheduled (which i dont think we DO handle very well XD https://github.com/tektoncd/pipeline/issues/2813) but i dont think that would be as relevant to your concerns b/c it would still be the same pod